### PR TITLE
Gear test lib

### LINKED ISSUE #570: Time-dependent tests in smart contracts
When testing a contract, a situation may arise,  where the action you are testing can only be tested after some time has passed.
It is necessary to add a feature to manipulate blockchain time to test time-dependent logic in smart contracts.

### LINKED ISSUE #571: Testing contracts with mocks
In complex contracts often we have to test just small parts of the contract. The testing of this small part may require a very specific contract state that in turn may depend on the state of other contracts. 
I think instead of establishing whole predefined smart contract system  it would be very convenient to create mock contracts. Mock contracts must be defined what they should reply for different actions and arguments and must behave very similar to the original ones.

### LINKED ISSUE #577: Create a tool for testing programs
### Problem to Solve

We need to have a tool for comprehensive testing and debugging programs. The best way is to use the cargo's test infrastructure.

### Possible Solution

Add a special crate and enable `cargo test` in programs.

- [ ] Make `cargo test` runnable (add implementations of syscalls, deal with the panic handler, etc)
- [ ] Add an MVP for a testing crate with the very basic utility functions (submit a program, send messages, view replies)

### PR DESCRIPTION
**Pull request by breathx**
Resolves #577. Closes #570 and closes #571 as not actual, because they're implemented in the lib.

**Release Notes**: *To be written.*

### CODE REVIEW

File: gear-test-lib/src/system.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    pub fn get_user(&self) -> ProgramId {
+        self.0.lock().unwrap().user
+    }
+
+    pub fn assert_log<E: Encode>(&self, from: u64, payload: E) {
+        self.assert_log_bytes(from, payload.encode())
+    }
+
+    pub fn assert_log_bytes<T: AsRef<[u8]>>(&self, from: u64, payload: T) {
```
Comment (NikVolf) [inline gear-test-lib/src/system.rs:75]: this should use `assert!` internally

File: examples/ping/Cargo.toml [hunk 2]
```diff
@@ -10,3 +10,6 @@ crate-type = ["cdylib"]
 
 [dependencies]
 gstd = { path = "../../gstd", features = ["debug"] }
+
+[dev-dependencies]
+gstd = { path = "../../gstd", features = ["debug"] }
```
Comment (shamilsan) [inline examples/ping/Cargo.toml:15]: Is it used?

File: gear-test-lib/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        let ping_pong = Program::from_file(
+            &sys,
+            "../target/wasm32-unknown-unknown/release/demo_ping.wasm",
+        );
+
+        ping_pong.send_bytes("INIT");
+        sys.assert_log_empty();
+
+        ping_pong.send_bytes("PING");
+        sys.assert_log_bytes(1, "PONG");
```
Comment (shamilsan) [inline gear-test-lib/src/lib.rs:43]: Similarly, better would be `assert_eq!(sys.log[1], b"PONG")`.

**Review by shamilsan — COMMENTED** [review]
Awesome as an MVP but needs to be reviewed while adopting.

File: gtest/src/manager.rs [hunk 4]
```diff
@@ -1,97 +1,249 @@
...
+        if !self.message_queue.is_empty() {
+            panic!("Message queue isn't empty");
         }
     }
+
+    fn init_success(&mut self, message_id: MessageId, program_id: ProgramId) {
+        self.programs
+            .get_mut(&program_id)
+            .expect("Can't find existing program")
+            .1 = ProgramState::Initialized;
```
Comment (gshep) [inline gtest/src/manager.rs:212]: Again for readability:
```
let (_, ref mut state, ..) = self.programs.get_mut(&program_id).expect(/* ... */);
state = ProgramState::Initialized;
```

File: gtest/src/manager.rs [hunk 5]
```diff
@@ -1,97 +1,249 @@
...
+                self.wake_message(message_id, program_id, id);
+            }
+        }
+    }
+
+    fn init_failure(&mut self, program_id: ProgramId) {
+        self.programs
+            .get_mut(&program_id)
+            .expect("Can't find existing program")
+            .1 = ProgramState::FailedInitialization
```
Comment (gshep) [inline gtest/src/manager.rs:225]: and use destructuring here too

File: gear-test-lib/src/system.rs (hunk 1 above)
Comment (breathx) [reply gear-test-lib/src/system.rs]: Now custom assertions removed. Only native ones in tests

File: examples/ping/Cargo.toml (hunk 2 above)
Comment (breathx) [reply examples/ping/Cargo.toml]: Nope. Removed

File: gear-test-lib/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        let sys = System::new();
+        sys.init_logger();
+
+        let ping_pong = Program::from_file(
+            &sys,
+            "../target/wasm32-unknown-unknown/release/demo_ping.wasm",
+        );
+
+        ping_pong.send_bytes("INIT");
+        sys.assert_log_empty();
```
[the remark this answers is not in the document]
Comment (breathx) [reply gear-test-lib/src/lib.rs]: That's it now

File: gear-test-lib/src/lib.rs (hunk 3 above)
Comment (breathx) [reply gear-test-lib/src/lib.rs]: Same as one above

File: gtest/src/manager.rs [hunk 7]
```diff
@@ -1,97 +1,249 @@
...
-impl ExecutionContext for InMemoryExtManager {
-    fn store_program(&self, program: gear_core::program::Program, _init_message_id: MessageId) {
-        self.waiting_init.borrow_mut().insert(program.id(), vec![]);
-        self.programs.borrow_mut().insert(program.id(), program);
+impl ExtManager {
+    pub(crate) fn new() -> Self {
+        Self {
+            msg_nonce: 1,
+            id_nonce: 1,
+            user: 100001.into(),
```
[the remark this answers is not in the document]
Comment (breathx) [reply gtest/src/manager.rs]: Good tip! Done

File: gtest/src/manager.rs (hunk 4 above)
Comment (breathx) [reply gtest/src/manager.rs]: Done with the same idea and a bit different syntax

File: gtest/src/manager.rs (hunk 5 above)
Comment (breathx) [reply gtest/src/manager.rs]: Same as one above

File: gtest/src/manager.rs [hunk 8]
```diff
@@ -1,97 +1,249 @@
...
+                .expect("Somehow message queue contains message for user");
+
+            if let ProgramState::FailedInitialization = state {
+                panic!(
+                    "Program with id {} failed it's initialization, so can't receive messages",
+                    message.dest(),
+                )
+            }
+
+            let kind = if let Some(kind) = message
```
[the remark this answers is not in the document]
Comment (breathx) [reply gtest/src/manager.rs:108]: How do you imagine it?

[the remark this answers is not in the document]
Comment (breathx) [reply gtest/src/manager.rs:108]: Just provide `get_entry(&Message) -> Option<EntryPoint>` for readability?

[the remark this answers is not in the document]
Comment (breathx) [reply gtest/src/manager.rs]: Simplified

**Review by LouiseMedova — APPROVED** [review]
It is a very handy tool that allows for total testing of the smart contract logic.
You can easily initialize the programs, send the messages and check the execution result in the log. The log can be checked for a specific message generated by the program execution and that's great since there may be several messages and it may be necessary to check not the last message, but some message that occurred during the program execution.
The tool makes it possible to increase the blockchain time and thus test the time-dependent logic in smart contracts. It also includes a helpful feature like creating mock contracts.

File: gtest/src/program.rs [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+            panic!("Invalid identifier: {:?}", other)
+        }
+
+        Self(bytes.into())
+    }
+}
+
+pub struct Program<'a> {
+    pub(crate) manager: &'a RefCell<ExtManager>,
+    pub(crate) id: ProgramId,
```
Comment (shamilsan) [inline gtest/src/program.rs:66]: Could we provide a method to get the ID of the program (when it has been assigned automatically)?

**Review by shamilsan — APPROVED** [review]
Excellent!

File: gtest/src/log.rs [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+    fn from(other: (ID, ID, T)) -> Self {
+        Self::builder()
+            .source(other.0)
+            .dest(other.1)
+            .payload_bytes(other.2)
+    }
+}
+
+impl Log {
+    pub fn builder() -> Self {
```
[the remark this answers is not in the document]
Comment (breathx) [reply gtest/src/log.rs:70]: Decided to left it so atm. `new()` isn't suit enough too, as I think

File: gtest/src/program.rs (hunk 9 above)
Comment (breathx) [reply gtest/src/program.rs:72]: Done `program.id() -> ProgramId`

**Comment by NikVolf** [comment]
> It is a very handy tool that allows for total testing of the smart contract logic.
> You can easily initialize the programs, send the messages and check the execution result in the log. The log can be checked for a specific message generated by the program execution and that's great since there may be several messages and it may be necessary to check not the last message, but some message that occurred during the program execution.
> The tool makes it possible to increase the blockchain time and thus test the time-dependent logic in smart contracts. It also includes a helpful feature like creating mock contracts.

this looks it worth putting as release notes @breathx

**Review by NikVolf — APPROVED** [review]

**Pull request merged**